### PR TITLE
Retainer

### PR DESCRIPTION
**Pull request by muXxer**

### CODE REVIEW

File: pkg/retainer/retainer/retainer.go [hunk 1]
```diff
@@ -18,93 +17,48 @@ import (
...
 type Retainer struct {
-	retainerFunc            RetainerFunc
-	blockFromCacheFunc      BlockFromCacheFunc
-	blockFromStorageFunc    BlockFromStorageFunc
+	store                   RetainerFunc
 	latestCommittedSlotFunc LatestCommittedSlotFunc
 	finalizedSlotFunc       FinalizedSlotFunc
 	errorHandler            func(error)
 
+	api     func(index iotago.SlotIndex) iotago.API
```
Comment (muXxer) [inline pkg/retainer/retainer/retainer.go:31]: currentAPIFunc

File: pkg/retainer/retainer/retainer.go [hunk 2]
```diff
@@ -157,143 +130,106 @@ func (r *Retainer) Shutdown() {
...
+	if err := r.store(blockID.Index()).StoreBlockFailure(blockID, failureCode); err != nil {
+		r.errorHandler(ierrors.Wrap(err, "failed to store block failure in retainer"))
+	}
 }
 
-func (r *Retainer) RetainTransactionFailure(transactionID iotago.TransactionID, slotIndex iotago.SlotIndex, failureCode apimodels.TransactionFailureReason) {
-	retainerStore := r.retainerFunc(slotIndex)
-	_ = retainerStore.StoreTransactionFailure(transactionID, failureCode)
+func (r *Retainer) RetainTransactionFailure(blockID iotago.BlockID, err error) {
+	failureCode := determineTxFailureReason(err)
```
Comment (muXxer) [inline pkg/retainer/retainer/retainer.go:159]: you could inline this in the next line

File: pkg/storage/prunable/retainer.go [hunk 3]
```diff
@@ -1,162 +1,183 @@
...
+	transactionStorePrefix
 )
 
+type BlockRetainerData struct {
+	State         apimodels.BlockState
+	FailureReason apimodels.BlockFailureReason
+}
+
+func (b *BlockRetainerData) Bytes() ([]byte, error) {
+	marshalUtil := marshalutil.New()
```
Comment (muXxer) [inline pkg/storage/prunable/retainer.go:23]: Can we initialize the util with the size already? Could reduce allocation operations.

File: pkg/storage/prunable/retainer.go [hunk 4]
```diff
@@ -1,162 +1,183 @@
...
+}
+
+func (b *BlockRetainerData) FromBytes(bytes []byte) (int, error) {
+	marshalUtil := marshalutil.New(bytes)
+	state, err := marshalUtil.ReadUint8()
+	if err != nil {
+		return 0, err
+	}
+	b.State = apimodels.BlockState(state)
+	reason, err := marshalUtil.ReadUint8()
```
Comment (muXxer) [inline pkg/storage/prunable/retainer.go:37]: I would add a newline here.

File: pkg/storage/prunable/retainer.go [hunk 5]
```diff
@@ -1,162 +1,183 @@
...
+	return marshalUtil.ReadOffset(), nil
+}
+
+type TransactionRetainerData struct {
+	State         apimodels.TransactionState
+	FailureReason apimodels.TransactionFailureReason
+}
+
+func (t *TransactionRetainerData) Bytes() ([]byte, error) {
+	marshalUtil := marshalutil.New()
```
Comment (muXxer) [inline pkg/storage/prunable/retainer.go:52]: Can we initialize the util with the size already? Could reduce allocation operations.

File: pkg/storage/prunable/retainer.go [hunk 6]
```diff
@@ -1,162 +1,183 @@
...
+}
+
+func (t *TransactionRetainerData) FromBytes(bytes []byte) (int, error) {
+	marshalUtil := marshalutil.New(bytes)
+	state, err := marshalUtil.ReadUint8()
+	if err != nil {
+		return 0, err
+	}
+	t.State = apimodels.TransactionState(state)
+	reason, err := marshalUtil.ReadUint8()
```
Comment (muXxer) [inline pkg/storage/prunable/retainer.go:66]: I would add a newline here.

File: pkg/retainer/retainer/errors.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	iotago "github.com/iotaledger/iota.go/v4"
+	"github.com/iotaledger/iota.go/v4/nodeclient/apimodels"
+)
+
+var txErrorsFailureReasonMap = map[error]apimodels.TransactionFailureReason{
+	// unknown type / type casting errors
+	iotago.ErrTxTypeInvalid:               apimodels.TxFailureTxTypeInvalid,
+	iotago.ErrUnknownInputType:            apimodels.TxFailureUTXOInputInvalid,
+	iotago.ErrUTXOInputInvalid:            apimodels.TxFailureUTXOInputInvalid,
+	iotago.ErrUnknownOutputType:           apimodels.TxFailureUTXOInputInvalid,
```
Comment (muXxer) [inline pkg/retainer/retainer/errors.go:14]: Mhh but its about the output... isn't that misleading?

**Review by daria305 — APPROVED** [review]

**Pull request merged**